### ISSUE TITLE
New transforms for building Looks

### ISSUE THREAD

**Issue by doug-walker** [AcademySoftwareFoundation/OpenColorIO#1100]
Today in OCIO, looks are generally built from either an ASC CDL, or a 3d-LUT. An ASC CDL is easily understood due to its parametric nature but is limited to fairly basic adjustments. A 3d-LUT allows for very powerful adjustments, but it’s difficult to understand, at a glance, what it is modifying. It would be useful to have some additional options that would occupy the middle ground.

This issue proposes a new set of transforms that provide some of the power of a 3d-LUT but with the parametric nature of an ASC CDL. This should allow for the creation of Look Transforms that are powerful and yet easy to understand. 

The first new transform is for making what are often termed “primary” color corrections to adjust overall exposure, color balance, and saturation. A second “tone” transform allows for more targeted adjustments to specific parts of the tonal range. For example, adjusting the brightness or color balance of just a portion of the highlights or shadows. The range over which each control operates is also parametrically adjustable. A third “curves” transform may be used to define spline-based RGB tonal curves using control points. All of the transforms work in both the forward and inverse directions.

Traditional color grading tools are often intended to work on specific types of color space encodings. For example, applications will sometimes offer separate tools for video vs. logarithmic color space encodings. This is because the aesthetic impact of a given mathematical operator (e.g., a gain/multiply) on an image varies greatly with how that image is encoded. The new tools offer three modes intended for video, logarithmic, and scene-linear color encodings. The algorithms have been carefully designed to avoid clamping, which is particularly important when working with wide-range scene-linear color encodings.

They would also leverage the dynamic property infrastructure introduced in PR #743 so that they may be created in real-time using a GPU (in situations where it is appropriate to create them).  Note: When this proposal was discussed at the OCIO v2 working group meeting, there was a lot of discussion about the fact that the transforms would use the dynamic properties and concerns over workflow problems caused by editing of configs.  

Just to be clear, the present issue is neither pro or con on the subject of editing transforms referenced by configs -- that is a side issue that is separate from the subject of adding new transforms.  After all, there is nothing that prevents the existing ASC CDL or Matrix Transforms already available from being easily edited.  All of these look transforms do need to be created at some point, and it is obviously important that that happen in the context of the rest of the viewing pipeline.

**Comment by KelSolaar** [comment]
Hi,

Isn't this a bit out of scope for OCIO? Where do we put the line between colour management and colour grading? This proposal and the related #1102 PR is very much stepping in the second domain.

What really troubles me (hence this comment) is that this is a rather important feature with important consequences and I could not, so far, find any references about it in the TSC Meetings, Github Issues (besides this very issue and the related PR) or Slack. There is also no way that @BernardLefebvre, as good developer as he is has produced over 20K lines of code in the 10-15 hours interval between this issue and the PR, implying that this feature has been built for quite some time behind the curtain and behind closed doors which seems orthogonal to the principles of openness. I'm then left wondering what is it serving primarily, OCIO or something else?

Don't get me wrong, I'm not criticising the feature in itself which could be useful, CLF 4? :), more the abrupt way it is brought to the table without apparent prior community discussion.

Cheers,

Thomas

**Comment by doug-walker** [comment]
Hi Thomas,

Thanks for your forthright feedback.  This proposal was discussed extensively at the OCIO v2 working group meeting on Aug. 6th.  (I will post the meeting minutes for that soon, it's been a busy time with a lot happening as we approach SIGGRAPH.)  We ran out of time at the previous working group meeting.  Also, when PR #743 was discussed at the OCIO v2 working group meeting last year, a number of people suggested that a feature like this would be very useful.

This issue #1100 was created as a place to capture follow-up conversations such as yours.  

I agree that there is some overlap between the creation of Looks to be used within color management and the task of color grading.  Often it is the same people that are involved in both.  However, as you know, color grading is a separate task from color management and requires specialized tools that go way beyond what OCIO is able to do (such as grading through shapes, animation, keying, etc.).

This feature is intended as a way of creating Looks that may be used within the existing OCIO architecture, and it is certainly not something that could be considered a color grading system.  There is already a precedent of other transforms (ASC CDL and LUT3D Transforms specifically) that are used to create Looks, so this feature should be considered an extension of those existing capabilities rather than as something entirely different.  Being able to replace a LUT3D with something more compact brings many benefits.

Regarding your statements about the development process, perhaps you should create a separate issue for that.  We have been doing our best to involve the community in our development of new features via the OCIO v2 working group.  However we are very open to your suggestions for how to improve on that.

Doug

**Comment by KelSolaar** [comment]
> This feature is intended as a way of creating Looks that may be used within the existing OCIO architecture, and it is certainly not something that could be considered a color grading system. There is already a precedent of other transforms (ASC CDL and LUT3D Transforms specifically) that are used to create Looks, so this feature should be considered an extension of those existing capabilities rather than as something entirely different.

Right, thanks for outlining the scope, it makes much much more sense and has me less worried :)

> We have been doing our best to involve the community in our development of new features via the OCIO v2 working group. However we are very open to your suggestions for how to improve on that.

Maybe it is a case for exposing a public roadmap, for example, Epic uses Trello to surface some of their largest features: https://trello.com/b/TTAVI7Ny/ue4-roadmap

Github has some similar capabilities nowadays, this would certainly be useful for those who cannot attend the Working Group meetings, missed one and when the info is not in the TSC notes!

**Comment by zachlewis** [comment]
I haven't had much of a chance to really dig in to these yet, but I did have a few questions and comments after briefly looking through the code and toying with the Python API on and off over the past couple of weeks --

1. These are really cool. I think you guys did a good job of designing these transforms to be as user- / author- / developer- friendly as possible, considering the complexity and flexibility. 
2. I've yet to see how the transforms "feel" in the various Log / Lin / Video modes, but I'm really looking forward to that.
3. Are GradingRGBCurveTransforms suitable for implementing, say, our favorite ACES (1.1+) Output Transform B-splines? If all one has is spline coefficients (i.e., the ACES 1.0 RRT, ODTs), that would seems like more of a challenge... 
4. These transforms really open up a lot of opportunities for building up expressive, creative analytic LMT libraries in ways that really aren't massively accessible today. I know LUT-authoring and creative grading and ACES aren't OCIO's focus, but it'll nevertheless be pretty cool to see more publicly available LMTs in the ether. Equally interesting opportunities for ALF2, IPP2, etc. of course. Just an observation.
5. I recall seeing references to HSL vs RGB curves...?
6. Any thoughts or developments re: preserving the "state" of dynamic parameters? If I had access to these kinds of adjustments in a viewer buffer, it would be really handy to have a way to save those adjustments.

**Comment by doug-walker** [comment]
Thanks for the feedback Zach.  I'll reply point by point:
1.  Thanks!
2.  Yes, the feel in the different modes was definitely something we worked on.
3.  That is a very perceptive question.  ;)   That was indeed the goal.  The splines I implemented here are a more advanced version of what I used for the original ACES CTL, so the representation is compatible with that purpose.  That is why we have not added any ACES Output Transforms to the Built-in Transforms yet, I was waiting for this to be merged.
4.  Yes.  One of the goals of ACES is that there is a fixed Output Transform and then people make tweaks to that as needed and store those as an LMT (which could become a Look Transform in OCIO).  That way it's modular and you know exactly what the look part is.  That is greatly preferred over people making their own monolithic output transforms that bake their look in with everything else.  One of the goals of this work was to help put some of the building blocks in place to encourage more and better use of look transforms that is friendlier to the rest of the pipeline.
5.  That could be an area for future development and we're open to collaboration on that.
6.  All of these parameters are already exposed in both Yaml and CTF, so they will already be saved the same was as anything else.  Whether they are dynamic or not doesn't really matter, it's no different than matrix coefficients or any other parameter.
